rfcs/20201027-modular-tensorflow-graph-c-api.md

```diff
@@ -0,0 +1,527 @@
+   
+  In plugin side, plugin authors should first deserialize `TF_Buffer` into a `plugin::GraphDef` object, and then transform the graph. To successfully deserialize the buffer, plugin authors must keep a copy of `graph.proto` in plugin, and make it exactly the same as that in proper side, along with all other proto files which help to build a graph. Here lists all files needed and objects generated:<a id="proto"></a>
+  - [attr_value.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/attr_value.proto): AttrValue, NameAttrList 
+  - [cost_graph.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/cost_graph.proto): CostGraphDef
+  - [function.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/function.proto): FunctionDefLibrary, FunctionDef, GradientDef
+  - [graph.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/graph.proto): GraphDef
+  - [node_def.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/node_def.proto): NodeDef
+  - [op_def.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/op_def.proto): OpDef, OpDeprecation, OpList
+  - [op_performance_data.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/grappler/costs/op_performance_data.proto): SessionInfo, OpInfo, NormalDistribution, LogNormalDistribution, OpPerformance, OpPerformanceList
+  - [resource_handle.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/resource_handle.proto): ResourceHandleProto
+  - [tensor.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/tensor.proto): TensorProto, VariantTensorDataProto
+  - [tensor_shape.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/tensor_shape.proto): TensorShapeProto
+  - [types.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/types.proto): DataType, SpecializedType
+  - [versions.proto](https://github.com/tensorflow/tensorflow/blob/r2.3/tensorflow/core/framework/versions.proto): VersionDef
+
+  After optimizing, plugin authors need to serialize the optimized `GraphDef` object into `TF_Buffer` as output.
+
+  Serialization function `MessageToBuffer` is already defined, desearization function `BufferToMessage` will be added in proper to convert plugin passed `TF_Buffer` back to proper's `GraphDef`. In plugin side, plugin authors need to define `P_MessageToBuffer` and `P_BufferToMessage`, which should be the same as those defined in proper, except the namespace of protobuf.
+
+  Proper:
+  ```cpp
+  Status MessageToBuffer(const tensorflow::protobuf::MessageLite& in, TF_Buffer* out);
+  Status BufferToMessage(const TF_Buffer* in, tensorflow::protobuf::MessageLite& out);
+  ```
+
+  Plugin:
+  ```cpp
+  Status P_MessageToBuffer(const plugin::protobuf::MessageLite& in, TF_Buffer* out);
```

Comment:
Just to make sure, `P_MessageToBuffer` and `P_BufferToMessage` are not a part of the API surface, right?

Reply:
Yes. I think the name needs to be changed, they are confusing, maybe "plugin::MessageToBuffer" is more clear. I will clarify it in RFC.